Assignment3/Q1.py

```diff
@@ -1,128 +1,125 @@
-            words: a list of strings, list of words that can be formulated from a character grid and are
-                   in a dictionary
-            path: list of integer tuples (similar to path in DFS search) we keep here an information about the vertices
-                  visited so far in a search, it prevents visiting the same cell in a grid twice in a single word search
 
         Returns:
-            a list of strings, list of words that can be form using characters in a grid, stating from start vertex
-            (corresponding to some character in a grid) that are in a dictionary.
+            it is a void function, no value is returned but the function modifies its word_list argument which
+            accumulates the words found in a grid.
     """
-    # Get coordinates of the starting vertex in a search.
+    # Get coordinates of the starting entry in a search.
     (i, j) = start
 
     # Add a corresponding character to the prefix 'accumulated' so far.
-    # NOTE: We are case insensitive in this solution
-    prefix = prefix + grid[i][j].upper()
+    prefix = prefix + grid[i][j]
 
-    # Add a vertex to a path to prevent using the same grid cell twice in a search for single word.
-    path = path + [start]
+    # Indicate that the entry has been visited by search.
+    visited[i][j] = True
 
     # Early terminate - cut off the search tree branch if the prefix so far is not a prefix in a dictionary.
-    if not dictionary.isPrefix(prefix):
-        return words
-
-    # Add a word to word list if it is a word in a dictionary.
-    if dictionary.isWord(prefix):
-        words = words + [prefix]
-
-    # Recursive call, DFS-like search but searches ALL paths from a vertex
-    for edge in graph[start]:
-        if edge not in path:
-            words = word_search_recursive(dictionary, grid, graph, edge, prefix, words, path)
-    return words
-
+    if dictionary.isPrefix(prefix):
+        # Add a word to word list if it is a word in a dictionary.
+        if dictionary.isWord(prefix):
+            # Do not allow duplicates in a list of words.
+            if prefix not in word_list:
+                word_list += [prefix]
+
+        # Get rows and columns of a grid.
+        rows = len(grid)
+        columns = len(grid[0])
+
+        # Two arrays to keep track of adjacent entries.
+        rows_adjacent = [-1, -1, -1, 0, 0, 1, 1, 1]
+        columns_adjacent = [-1, 0, 1, -1, 1, -1, 0, 1]
+
+        # Recursive call, DFS-like search but searches ALL paths from an entry
+        # Iterating through the two arrays above and using 'is_not_visited' method all
+        # adjacent entries are found.
+        for k in range(8):
+            adjacent = (i+rows_adjacent[k], j+columns_adjacent[k])
+            if is_not_visited(adjacent[0], adjacent[1], rows, columns, visited):
+                word_search_recursive(dictionary, grid, visited, adjacent, word_list, prefix)
```

Comment:
When using a keyword argument (the optional argument `prefix` in this case), it is a useful convention to always mention it by name, i.e., in this case:
```python
    word_search_recursive(dictionary, grid, visited, adjacent, word_list, prefix=prefix)
```
What you did is not incorrect, but using the explicit name helps if you add a second optional argument (or change the order of optional arguments), as things will always work correctly with the named argument. The current implementation depends on the fact that prefix is the first optional argument.